File: ios/Ausweisapp2Wrapper.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(Ausweisapp2Wrapper, RCTEventEmitter <RCTInvalidating>)

RCT_EXTERN_METHOD(start)
RCT_EXTERN_METHOD(stop)
RCT_EXTERN_METHOD(send: (NSString *) command)
RCT_EXTERN_METHOD(isRunning: (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)
RCT_EXTERN_METHOD(supportedEvents)

-(void)startObserving {}

-(void)stopObserving {}

+ (BOOL)requiresMainQueueSetup
{
  return NO;
}

@end

File: ios/Ausweisapp2Wrapper.swift
import AusweisApp
import Foundation

@objc(Ausweisapp2Wrapper)
class Ausweisapp2Wrapper: RCTEventEmitter {
  static var shared: Ausweisapp2Wrapper?

  override init() {
    super.init()
    Ausweisapp2Wrapper.shared = self
  }

  /**
   * Callback that handles Messages sent by the AusweisApp SDK
   */
  private var aaCallback: AusweisAppCallback = { (msg: UnsafePointer<CChar>?) in
    guard let msg: UnsafePointer<CChar> = msg else {
      Ausweisapp2Wrapper.shared?.sendEvent(withName: "connected", body: nil)
      return
    }

    let message = String(cString: msg)
    Ausweisapp2Wrapper.shared?.sendEvent(withName: "message", body: message)
  }

  /**
   * Initialize the AusweisApp SDK
   */
  @objc func start() {
    if !ausweisapp_init(aaCallback, nil) {
      Ausweisapp2Wrapper.shared?.sendEvent(withName: "error", body: nil)
    }
  }

  /**
   * Stop the AusweisApp SDK (Cancels all running Workflows)
   */
  @objc func stop() {
    ausweisapp_shutdown()
    Ausweisapp2Wrapper.shared?.sendEvent(withName: "disconnected", body: nil)
  }

  /**
   * Send a command to the AusweisApp SDK encoded as a JSON String
   */
  @objc func send(_ command: String) {
    ausweisapp_send(command)
  }

  /**
   * Return if AusweisApp SDK is running
   */
  @objc func isRunning(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(ausweisapp_is_running())
  }

  override func invalidate() {
    stop()
  }

  /**
   * Supported event types by the NativeEventEmitter
   */
  override func supportedEvents() -> [String]! {
    return ["message", "connected", "error", "disconnected"]
  }
}
